engine/src/gas.rs

```diff
@@ -0,0 +1,563 @@
+        match deconstruct_storage_key(key) {
+            None => self.inner.read_storage(key),
+            Some((address, index)) => match self.state_override.get(&address) {
+                None => self.inner.read_storage(key),
+                Some(state_override) => state_override
+                    .get(&index)
+                    .map(|value| EngineStorageValue::Vec(value.as_bytes().to_vec())),
+            },
+        }
+    }
+
+    fn storage_has_key(&self, key: &[u8]) -> bool {
+        self.read_storage(key).is_some()
+    }
+
+    fn write_storage(&mut self, key: &[u8], value: &[u8]) -> Option<Self::StorageValue> {
+        self.inner.write_storage(key, value)
+    }
+
+    fn write_storage_direct(
+        &mut self,
+        key: &[u8],
+        value: Self::StorageValue,
+    ) -> Option<Self::StorageValue> {
+        self.inner.write_storage_direct(key, value)
+    }
+
+    fn remove_storage(&mut self, key: &[u8]) -> Option<Self::StorageValue> {
+        self.inner.remove_storage(key)
+    }
+}
+
+const STORAGE_VERSION: u8 = storage::VersionPrefix::V1 as u8;
+const STORAGE_PREFIX: u8 = storage::KeyPrefix::Storage as u8;
+
+fn deconstruct_storage_key(key: &[u8]) -> Option<(H160, H256)> {
+    let version = *key.first()?;
+    if version != STORAGE_VERSION {
+        panic!("Unexpected version");
+    }
+    if key.get(1)? == &STORAGE_PREFIX {
+        let key_len = key.len();
+        // Lengths are 54 or 58 bytes, depending on if the generation is present or not
+        if key_len == 54 {
+            let address = H160::from_slice(&key[2..22]);
+            let value = H256::from_slice(&key[22..54]);
+            Some((address, value))
+        } else if key_len == 58 {
+            let address = H160::from_slice(&key[2..22]);
+            let value = H256::from_slice(&key[26..58]);
+            Some((address, value))
+        } else {
+            panic!("Unexpected storage key length")
+        }
+    } else {
+        None
+    }
+}
+
+#[derive(Debug, serde::Serialize)]
+#[serde(tag = "request_nonce_type")]
+pub enum NonceStatus {
+    NotProvided { current_nonce: u64 },
+    TooLow,
+    GreaterOrEqual { current_nonce: u64 },
+}
+
+#[derive(Debug, serde::Serialize)]
+pub enum StateOrEngineError {
+    StateMissing,
+    Engine(EngineError),
+}
+
+#[allow(clippy::too_many_arguments)]
+fn compute_call_result<I: aurora_engine_sdk::io::IO + Copy>(
+    io: I,
+    from: Address,
+    to: Option<Address>,
+    gas_limit: u64,
+    gas_price: U256,
+    value: Wei,
+    data: Vec<u8>,
+    nonce: Option<u64>,
+    env: aurora_engine_sdk::env::Fixed,
+) -> Result<SubmitResult, StateOrEngineError> {
+    let mut handler = aurora_engine_sdk::promise::Noop;
+    aurora_engine::state::get_state(&io)
+        .map_err(|_| StateOrEngineError::StateMissing)
+        .and_then(|engine_state| {
+            let mut engine = aurora_engine::engine::Engine::new_with_state(
```

Comment:
Import `Engine` at the top along with `aurora_engine_modexp::AuroraModExp` and then can write

```
let mut engine: Enigne<_, _, AuroraModExp> =Engine::new_with_state(
```